Replace legacy faucet with code from aptos-tap

banool opened the pull request:
## Summary
This PR replaces the legacy faucet with the new faucet from aptos-tap. This is not entirely just a drop-in replacement, I've done some refactoring of the new faucet code and the old faucet CLI to make them work with other pieces in aptos-core like the Aptos CLI and the smoke tests.

This change only actually changes the CLI and the smoke tests, the prod faucets are still using aptos-tap, I'll cut them over later.

To avoid bringing in the SeaORM deps in the standard case I've hidden that Checker behind a feature flag.

I removed any reference to any internal doc / repo. Accordingly some of the READMEs are now a bit lacking, I'll fix those up later.

I have changed the license from All Rights Reserved to Apache 2.0.

## Testing
Testing the CLI:
```
cargo run -- --amount 5 --accounts 0xb078d693856a65401d492f99ca0d6a29a0c5c0e371bc2521570a86e40d95f823 --key <key> --mint-account-address 0xa550c18
```

Testing the Aptos CLI with a faucet:
```
cargo run -p aptos -- node run-local-testnet --force-restart --assume-yes --with-faucet
```

Running the SDK tests against the local testnet + faucet:
```
pnpm test
```

The faucet tests themselves:
```
$ cargo test -p aptos-faucet-core
running 6 tests
test server::run::test::test_transfer_health ... ignored
test server::run::test::test_bypassers ... ok
test server::run::test::test_redis_ratelimiter ... ok
test server::run::test::test_checkers ... ok
test server::run::test::test_mint_funder_wait_for_txns ... ok
test server::run::test::test_mint_funder ... ok
```

Testing docker/compose/validator-testnet/docker-compose.yaml (on WSL on an x86 machine). I changed `aptoslabs` to `aptos-core` in the yaml so it would use the images I built locally.
```
./docker/docker-bake-rust-all.sh
cd docker/compose/validator-testnet
IMAGE_TAG=from-local docker-compose up
```

clay-aptos commented on crates/aptos-faucet/ts-client/CHANGELOG.md:
```diff
@@ -0,0 +1,14 @@
+
+All notable changes to the client will be captured in this file. This changelog is written by hand for now. It adheres to the format set out by [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
+
+**Note:** The client does not follow semantic version while we are in active development. Instead, breaking changes will be announced with each devnet cut. Once we launch our mainnet, the SDK will follow semantic versioning closely.
```
Should this be updated? "Once we launch our mainnet"

banool replied:
Ah yes, good call out.

clay-aptos reviewed (commented):
Thanks, Daniel. The docs here LGTM with some small changes I made directly and one comment I left for you. I will let Greg review before approving.

Please note, I now link to:
https://github.com/aptos-labs/aptos-core/tree/main/crates/aptos-faucet

From:
https://aptos.dev/guides/get-test-funds

I hope that makes sense. I like the descriptions of subdirectories you added.

gregnazario commented on crates/aptos-faucet/core/README.md:
```diff
@@ -0,0 +1 @@
+# Aptos Tap
```
I think you can remove this file if it doesn't describe how to run it anymore.  Though, I think we need instructions before removing the previous faucet

banool replied:
Yeah I mentioned in summary that I removed the README for now because it referred to internal docs, I'll fix that up.

gregnazario commented on crates/aptos-faucet/core/src/bypasser/auth_token.rs:
```diff
@@ -0,0 +1,42 @@
+// Copyright © Aptos Foundation
```
You'll need to add the appropriate full Apache license header, not just the copyright line.

banool replied:
Yep just did that, spoke with Josh to confirm which to do for what files.

gregnazario commented on crates/aptos-faucet/ts-client/package.json:
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "aptos-tap-client",
+  "description": "Generated client for the Aptos Tap",
+  "license": "UNLICENSED",
```
This should now be Apache

gregnazario commented on crates/aptos-faucet/ts-client/CHANGELOG.md:
```diff
@@ -0,0 +1,14 @@
+
+All notable changes to the client will be captured in this file. This changelog is written by hand for now. It adheres to the format set out by [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
+
+**Note:** The client does not follow semantic version while we are in active development. Instead, breaking changes will be announced with each devnet cut. Once we launch our mainnet, the SDK will follow semantic versioning closely.
```
You can probably just remove this altogether

gregnazario commented on crates/aptos-faucet/migration/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+
+[dependencies]
+async-std = { version = "^1", features = ["attributes", "tokio1"] }
+sea-orm-migration = { version = "0.11.0", features = ["sqlx-postgres", "runtime-tokio-rustls"] }
```
Is there a reason why we're using yet another ORM for this rather than using diesel?

banool replied:
Because Max / Bowen recommended I try this instead of Diesel and I indeed found it better. Because of the feature this won't get pulled in the normal case, so it isn't too painful to have.

The SQL based checker is not actually used in prod btw, but you never know if we'll want it again at some point and it's perfectly functional code, so I figure it's worth keeping (behind a feature).

gregnazario commented on crates/aptos-faucet/metrics-server/README.md:
```diff
@@ -0,0 +1 @@
+Note: Much of this is yanked from https://github.com/aptos-labs/aptos-core/tree/main/crates/inspection-service. Unfortunately adding a dependency on that meant bringing in the whole Aptos codebase, including stuff that would require me to enable tokio unstable (not a huge deal, but just an indicator of how much extra code I'd be bringing in). So this is a stripped down, more tokio native version  ...
```
Probably can remove this, or use the inspection service

banool replied:
I don't want to make more major changes in this PR, I'm trying to make it as much of a drop in as possible. We can cut over in a later PR. Though tbh this metrics server I have is better for how we use it here, the other is needlessly complex / heavy.

banool replied:
Also the inspection service is coupled pretty heavily to a node, it assumes you'll have a NodeConfig and whatnot.

gregnazario reviewed (commented):
There's a lot here, and I've been trying to follow it all.  Main thoughts:

This is a huge PR, it's very hard to keep track of.  How does one run the faucet now, before it was a single command and pretty straightforward, I'm not sure I found the readme or docs for that.

banool commented:
So as for how to run it, I'll add back the README, I just removed it for now bc it referred to too many internal docs.

In this PR I add a command to make it easy to run for some of the swarm local testnet cases and I updated the relevant READMEs accordingly.

banool commented:
Okay, wrote up a new, condensed README that just contains all the good stuff. I also fixed up the tests to get them working in aptos-core.

davidiw reviewed (commented):
This is effectively one massive PR that makes it rather unwieldy. How hard would it be to break it up into a few pieces?

* Pull in the core code and swap with the existing faucet
* Bring in helper scripts
* Bring SDKs (shouldn't those be under ecosystem?)
* Move code duplicates `aptos-move/move-examples/sources/minter.move`

Whatever we can do to cut down on the load of a massive 15,000 review would be great.

Remind me, we use the current faucet for local testing, but deployments we're using the new faucet? Considering that the new faucet is 10x bigger, is it worth evaluating if the old faucet is sufficient? My thinking is that this is a non-trivial service that might need support in the future, the simpler the better. Of course, stability and reliability matter a lot.

Maybe worth having a tech review (educational session?) on this, so we can warm up to it. That way we can ask our dumb questions :P

banool commented:
Yeah this is indeed a huge PR. If this were another place I'd do this with stacked PRs but... 

- So, the first point about the core code and swapping with the existing faucet is like 95% of this PR, so any other splitting won't make much of a dent in this PR. So if we want this PR to be smaller, I'll have to artificially cut up the core and make pseudo stacked PRs.
- Helper scripts, do you mean the SQL stuff?
- Yeah sure I could do the "SDK" (just a client generated from the OpenAPI spec) later. That'd probably be ideal since currently we can't use it fully because it refers to endpoints that don't exist with the old faucet.
- Oop I didn't realize minter.move was already in the repo.

As for the size of the review, yeah, not ideal. The code was ostensibly reviewed as it landed in aptos-tap, but in reality, not so much...

Correct, the current faucet exists only for local testing and use in the CLI local testnet. My thoughts are having two faucets leaves open the chance for desync in functionality, doubles the maintenance cost, increases deployment complexity, etc. The new faucet appears more complex yes but all the complexity can be disabled at runtime as part of configuration. The changes in running it both from the CLI and from testing are pretty minimal. For all the use cases here (CLI, tests, etc) all the fancy features are turned off and it's really just a minter behind an endpoint.

Tech review, definitely. I'll book something. Mostly I'll just go through the production readiness doc, which I've kept up to date.

banool commented:
Okay I removed the following from this PR:
- Move scripts
- TS client
- PostgresRatelimitChecker, which removes all the ORM stuff. We don't use this in prod anyway, and it's unlikely we will down the line.

This halves the number of directories, it should make it simpler to approach.

I also fixed up the tests to mark them as integration tests.

Still working on fixing the smoke tests.

clay-aptos commented on crates/aptos-faucet/DEV.md:
````diff
@@ -0,0 +1,130 @@
+curl -H 'Content-Type: application/json' -d '{"amount": 5, "address": "3c769ea16f38fdc218341c63ff8c1c5c7dcbb4d5d850675e92b09997fd36e8f0"}' -H 'CAPTCHA_KEY: 814861163' -H 'CAPTCHA_VALUE: Es9XE' -H 'what_wallet_my_guy: petra_of_course' localhost:10212/fund
+```
+
+See how this also includes headers to say "yessir I'm Petra".
````
Maybe replace "Yessir" here? :-)

clay-aptos commented on crates/aptos-faucet/README.md:
```diff
@@ -1,56 +1,107 @@
-* For new accounts as defined by the pub_key, the service first issues a transaction for creating the account and another for transferring funds.
-* All funds transferred come from the account 0xa550c18.
-* Clients should retry their request if the requests or the transaction execution failed. One reason for failure is that, under load, the service may issue transactions with duplicate sequence numbers. Only one of those transactions will be executed, the rest will fail.
+You can find many examples of different config files in `configs/`.
```
Can we link to the dir?

clay-aptos commented on crates/aptos-faucet/README.md:
````diff
@@ -1,56 +1,107 @@
+yarn generate-client
 ```
+
+## Internal
````
I highly recommend we remove the Internal section at the bottom!

banool replied:
Well it's handy for us to have these backlinks but sure 😛

clay-aptos reviewed (commented):
Thanks, Daniel. I edited the Markdown file changes in direct commits and left a few comments/suggestions.

davidiw reviewed (approved):
unblocking...


please consider using the cleaner paradigm for dispatch.

I think at some point it would be great to really analyze this and see what else we can do to minimize the code bloat or get the code to be a bit tighter...

my original thought for how this would be was to have a single core component that dispatched transactions and then modules for each of the different behaviors. It seems that's kinda what is here, but even the core component is spread across and then we have things like a CLI -- which is interesting, but seemingly undocumented and I don't see it as likely to be used.

gregnazario reviewed (approved):
Agreed, in the future we should probably simplify this, unblocking this.

banool commented:
@davidiw Yeah let's chat offline, I'd love to chat about how we can make this cleaner. In my mind besides this dispatch problem (I haven't seen this `enum_dispatch` crate before, very handy) I think the current structure is pretty good.

As for the CLI, I didn't have one originally, I've included this because we already had a CLI for the old faucet and I didn't want to break anyone's workflow.

banool commented:
The only failing test here is due to some weird CI feature where the old workflow description is still running. I'll force land this when I'm at my computer for enough time to make sure nothing explodes.

brmataptos commented on aptos-move/move-examples/scripts/minter/build/Minter/bytecode_scripts/main.mv:
Was this a spurious file that was checked in?  The location is pretty odd.

banool replied:
It was intentional, the faucet uses this compiled script.
